### PR TITLE
First-pass of OSD recording

### PR DESCRIPTION
**Pull request by Knifa**
This records OSD frames on every MSP draw screen message to `DJIxxxx.osd` alongside the `DJIxxxx.mp4` files. The entire character map gets written as a frame and the renderer ([osd-dump-tools](https://github.com/Knifa/osd-dump-tools)) is responsible for dealing with that currently.

Frames are synced to the last frame by hooking `duss_osal_task_create` to steal a `gs_lv_transcode_t` from `record_thread`. `lv_transcode->last_frame_idx` gives us the last frame index.

The `.osd` file starts with a header:

| Field | Value | Notes |
| --- | --- | --- |
| `magic` | `MSPOSD` (inc. terminator) | |
| `version` | `1` | Because this structure will definitely change! |
| `char_width` | `current_display->char_width` | |
| `char_height` | `current_display->char_width` | |
| `font_width` | `current_display->font_width` | |
| `font_height` | `current_display->font_width` | |
| `x_offset` | `current_display->x_offset` | |
| `y_offset` | `current_display->y_offset` | |
| `font_variant` | constant representing font variant | So the renderer can use the right font automagically. |

Then each frame:

| Field | Value | Notes |
| --- | --- | --- |
| `frame_idx` | `rec_get_frame_idx()` | For syncing frames |
| `size` | `[size of frame in bytes]` | |
| `data` | `msp_character_map` or `msp_render_character_map` | |

Caveats right now I can think of:
- After some time the transcoder will swap to a new file which kicks it out of `RECORD_STATE_RECORDING` for a moment --- that'll cancel OSD recording, right now.
- Absolutely no checking on SD card space, etc.
- Probably a better way of storing if it's HD or not, etc? I'm inclined to let the renderer handle it and store things as generally as possible.

### CODE REVIEW

**Comment by benlumley** [comment]
Nice work.

Format wise - headers should include the font size too? As in the pixel size of each box? And the X and y offset... So everything in the display info struct...

Think then it's flexible/future proof

File: jni/rec/rec_shim.c [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+extern gs_lv_transcode_t *rec_lv_transcode;
+
+duss_osal_priority_t duss_osal_task_create(
+    duss_osal_task_attrib_t *task_attrib,
+    duss_osal_task_handle_t **task_handle)
+{
+  if (duss_osal_task_create_orig == NULL)
+  {
+    duss_osal_task_create_orig = dlsym(RTLD_NEXT, "duss_osal_task_create");
```
Comment (bri3d) [inline jni/rec/rec_shim.c:25]: Has this been reliable for you? I've had issues with RTLD_NEXT returning NULL and having to load the lib manually for some libs. If this is reliable though it does save a lot of jankiness

**Comment by bri3d** [comment]
Agree with Ben re: the size/offset stuff, avoids pain down the road when people add 10,000 new OSD layout formats like they seem to be chomping at the bit to do. 

Otherwise no complaints, looks pretty clean and straightforward. Including the full dumped header to get one struct is probably overkill but I'm not mad about it. 

Nicely done!

File: jni/rec/rec_shim.c (hunk 1 above)
Comment (Knifa) [reply jni/rec/rec_shim.c:25]: No problems here, at least with this function!

**Comment by Knifa** [comment]
Added the rest of the display info so I think we're good!

**Comment by mavenius** [comment]
:worksonmymachine: (for 16-odd seconds, anyway)

**Comment by bri3d** [comment]
The only other feedback I have, which is non-blocking, is that technically using `fwrite` in a critical path like this could block and freeze the OSD display, but it's very unlikely as the OS will buffer the writes by default. 

I do think this will crash if the SD card is removed during recording, but the process will get restarted so I'm also not too massively concerned about that.

I went down this rabbit hole with normal recording and the two solutions I found were to use async writes using the linux "aio" functionality, or threading 🤮 .

Anyway I think this is just a theoretical concern and not a practical one - if it's working well, I don't have a major opposition to just shipping it.

**Pull request merged**